### PR TITLE
Fix #115: JSON encoding in client API

### LINKED ISSUE #115: Client API JSON serialization of commands is broken
The client API uses the `DefaultJsonCodec` to marshall objects such as `CCommand`s, which doesn't produce the correct format.  Two problems are that the `EEnumLiteral` literal value is not used for the command type, instead it is the Java `enum` constant name; and cross-references are expressed as the expanded object definition, not an emfjson-jackson proxy.

For example, an add command looks like this:

```json
{
    "type": "ADD",
    "owner": {
        "name": null,
        "nodes": [],
        "flows": []
    },
    "feature": "nodes",
    "indices": [],
    "dataValues": [],
    "objectValues": [
        {
            "name": null,
            "duration": 0,
            "component": null
        }
    ],
    "objectsToAdd": [
        {
            "name": null,
            "duration": 0,
            "component": null
        }
    ]
}
```

instead of this:

```json
{
    "type": "add",
    "owner": {
        "eClass": "http://www.eclipsesource.com/modelserver/example/coffeemodel#//Workflow",
        "$ref": "sb3k.json#//@workflows.0",
    },
    "feature": "nodes",
    "objectValues": [
        {
            "$ref": "//@objectsToAdd.0",
        }
    ],
    "objectsToAdd": [
        {
            "eClass": "http://www.eclipsesource.com/modelserver/example/coffeemodel#//AutomaticTask",
            "name": "Brew"
        }
    ]
}
```

### PR DESCRIPTION
**Pull request by cdamus**
- use *emfjson-jackson* mapper in the `DefaultJsonCodec`
- ensure IDREFs for internal cross-referencees by using a temporary resource when marshalling
- make safe copies of commands in the server when shipping them around to various components, so that the notification doesn’t end up with an empty list of `objectsToAdd`

### CODE REVIEW

File: com.eclipsesource.modelserver.client/src/main/java/com/eclipsesource/modelserver/client/ModelServerClient.java [hunk 1]
```diff
@@ -283,6 +286,13 @@ private boolean isSupportedFormat(String format) {
     
 	@Override
 	public CompletableFuture<Response<Boolean>> edit(String modelUri, CCommand command, String _format) {
+		// Encapsulate the command in a resource before marshalling it so that internal
+		// cross-references are serialized as IDREFs (e.g., "//@objectsToAdd.0") instead of
+		// HREFs (e.g., "#//@objectsToAdd.0") which will not resolve in the Model Server
+		// on account of the resource with URI "" not existing
+        Resource cmdRes = new ResourceImpl(URI.createURI("$command.json"));
+        cmdRes.getContents().add(EcoreUtil.copy(command));
```
Comment (edgarmueller) [inline com.eclipsesource.modelserver.client/src/main/java/com/eclipsesource/modelserver/client/ModelServerClient.java]: Could we move this bit into the encoder? Formatting seems off

File: com.eclipsesource.modelserver.client/src/test/java/com/eclipsesource/modelserver/client/ModelServerClientTest.java [hunk 2]
```diff
@@ -229,15 +232,24 @@ public void configure() throws MalformedURLException, ExecutionException, Interr
...
-    	set.setFeature("name");
-    	set.getDataValues().add("Foo");
+    	Workflow flow = CoffeeFactory.eINSTANCE.createWorkflow();
+    	((InternalEObject)flow).eSetProxyURI(URI.createURI("SuperBrewer3000.json#//workflows.0"));
+    	CCommand add = CCommandFactory.eINSTANCE.createCommand();
+    	add.setType(CommandKind.ADD);
+    	add.setOwner(flow);
+    	add.setFeature("nodes");
+    	add.getObjectsToAdd().add(CoffeeFactory.eINSTANCE.createAutomaticTask());
+    	add.getObjectValues().addAll(add.getObjectsToAdd());
```
Comment (edgarmueller) [inline com.eclipsesource.modelserver.client/src/test/java/com/eclipsesource/modelserver/client/ModelServerClientTest.java]: Formatting

File: com.eclipsesource.modelserver.emf/src/test/java/com/eclipsesource/modelserver/emf/common/ModelControllerTest.java [hunk 3]
```diff
@@ -169,4 +174,39 @@ public void executeCommand() throws EncodingException, DecodingException {
...
+        queryParams.put("modeluri", Collections.singletonList("SuperBrewer3000.json"));
+        when(context.queryParamMap()).thenReturn(queryParams);
+        when(context.body()).thenReturn(commandAsString);
+        when(modelRepository.getResourceSet()).thenReturn(rset);
+        when(modelRepository.getModel("SuperBrewer3000.json")).thenReturn(Optional.of(task));
+        modelController.executeCommand(context, "SuperBrewer3000.json");
+
+        // Unload the resource so that the "owner" is a proxy in
+        // the expected command as well as the actual
+        res.unload();
```
Comment (edgarmueller) [inline com.eclipsesource.modelserver.emf/src/test/java/com/eclipsesource/modelserver/emf/common/ModelControllerTest.java]: Is this still needed?

Comment (cdamus) [reply com.eclipsesource.modelserver.emf/src/test/java/com/eclipsesource/modelserver/emf/common/ModelControllerTest.java]: No, indeed.

**Review by edgarmueller — APPROVED** [review]

**Pull request merged**